docs/source/rbac/tech-implementation.md

```diff
@@ -0,0 +1,49 @@
+# Technical Implementation
+
+Roles are stored in the database similarly as users, services, etc., and can be added or modified as explained in {ref}`define_role_target`. Objects can gain, change and lose roles. For example, one can change a token's role, and as such its permissions, without the need to initiate new token (currently through `update_roles` and `remove_obj` functions in `roles.py`, this will be eventually available through APIs). Roles' and scopes' utilities can be found in `roles.py` and `scopes.py` modules.
+
+## Resolving roles and scopes
+Roles and scopes are resolved on several occasions, for example when requesting an API token with specific roles or making an API request. The following sections provide more details.  
+
+### Requesting API token with specific roles
+API tokens grant access to JupyterHub's APIs. The RBAC framework allows for requesting tokens with specific existing roles. To date, it is only possible to add roles to a token in two ways: 
+1. through the `config.py` file as described in the {ref}`define_role_target` section 
+2. through the POST /users/{name}/tokens API where the roles can be specified in the token parameters body (see [](../reference/rest-api.rst)).
+
+The RBAC framework adds several steps into the token issue flow.
+
+If no roles are requested, the token is issued with a default role (providing the requester is allowed to create the token).
+
+If the token is requested with a specific role or multiple roles, the permissions of the would-be token's owner are checked against the requested permissions to ensure the token will not grant its owner additional privileges. In practice, this corresponds to resolving all the token owner's roles (including the roles associated with their groups) and the token's requested roles into a set of scopes. The two sets are compared and if the token's scopes are a subset of the token owner's scopes, the token is issued with the requested roles.
+
```

Comment:
I think this is the wrong place for my comment, but I did not find the origin pull request for this feature.

Granting a token access to arbitrary roles matching the users and groups scopes is dangerous, as user roles and thus scopes might change, therefore I think it will be wise to add a relation to users to tokens, and compare the scopes at runtime.

Reply:
Very good point, thank you. 

We realized that the token owner's roles/scopes can change after issuing tokens. Additional permissions check is implemented when the token is passed to requests. Instead of assessing only the token's scopes, the owner's scopes are checked again to make sure that the owner can still use the token (ref: #3363). 

The docs will provide more info about that as well.

Reply:
Yes, token scopes are always evaluated to the *intersection* of scopes on the token and those on the owner at request time, to ensure that a token never has more permissions than its owner if the owner's permissions are ever reduced in the future.

This paragraph only notes that any *attempt* to expand permissions at token creation time will be met with an error, rather than a token that silently doesn't actually have the requested permissions due to a limitation on the owner.

A clarification of that fact would be appropriate.

An `all` metascope preserves the current default behavior of "this token has all its owner's permissions" evaluated at request time.